src/solvers/flattening/arrays.cpp

```diff
@@ -825,9 +853,75 @@ void arrayst::add_array_constraints_if(
       or_exprt(literal_exprt(cond_lit),
       equal_exprt(index_expr1, index_expr2)));
     add_array_constraint(lazy, false); // added immediately
+    array_constraints_map[constraint_typet::ARRAY_IF_FALSE].push_back(
+      lazy.lazy);
 
 #if 0 // old code for adding, not significantly faster
     prop.lcnf(cond_lit, convert(equal_exprt(index_expr1, index_expr2)));
 #endif
   }
 }
+
+std::string arrayst::enum_to_string(constraint_typet type)
+{
+  switch(type)
+  {
+  case constraint_typet::ARRAY_ACKERMANN:
+    return "arrayAckermann";
+  case constraint_typet::ARRAY_EQUALITY:
+    return "arrayEquality";
+  case constraint_typet::ARRAY_WITH:
+    return "arrayWith";
+  case constraint_typet::ARRAY_WITH_OTHER:
+    return "arrayWithOther";
+  case constraint_typet::ARRAY_IF_TRUE:
+    return "arrayIfTrue";
+  case constraint_typet::ARRAY_IF_FALSE:
+    return "arrayIfFalse";
+  case constraint_typet::ARRAY_OF:
+    return "arrayOf";
+  case constraint_typet::ARRAY_TYPECAST:
+    return "arrayTypecast";
+  case constraint_typet::ARRAY_CONSTANT:
+    return "arrayConstant";
+  case constraint_typet::ARRAY_NON_CONSTANT:
+    return "arrayNonConstant";
+  case constraint_typet::ARRAY_COMPREHENSION:
+    return "arrayComprehension";
+  default:
+    UNREACHABLE;
+  }
+}
+
+void arrayst::display_constraints()
+{
+  json_objectt json_result;
+  json_objectt &json_array_theory =
+    json_result["arrayConstraints"].make_object();
+
+  size_t num_constraints = 0;
+
+  array_constraints_mapt::iterator it = array_constraints_map.begin();
+  while(it != array_constraints_map.end())
+  {
+    std::string contraint_type_string = enum_to_string(it->first);
+    json_arrayt &json_constraint_list =
+      json_array_theory[contraint_type_string].make_array();
+
+    array_constraintst::iterator constraint_it = it->second.begin();
+    while(constraint_it != it->second.end())
+    {
+      std::stringstream ss;
+      ss << format(*constraint_it);
+      json_constraint_list.push_back(json_stringt(ss.str()));
+
+      constraint_it++;
+      num_constraints++;
+    }
+    it++;
+  }
+
+  json_result["numOfConstraints"] =
+    json_numbert(std::to_string(num_constraints));
+  log.status() << ",\n" << json_result;
+}
```

Comment:
What is an elegant way to exit CBMC once the display is done? Adding `exit(CPROVER_EXIT_SUCCESS)` here makes the json output messy.

Reply:
You must not exit directly from here. This code may be called in a number of ways, and it will be up to the caller to exit in a clean manner. As far as I understand, currently `post_process()` is the only place calling this. Now I'd question whether this should really be called _from_ `post_process()`, or rather be called from some place _after_ `post_process()` has been called.

Reply:
Noted.

Reply:
If you want to terminate without sat-solving then the intended mechanism is to do that by using the symex-only checkers. Those run the pipeline until after symex and then terminate.
You could possibly do that in the `src/goto-checker/multi_path_symex_only_checker.cpp` which is used to handles similar cases such as `--show-vcc`, `--show-program`, etc.

Reply:
The array constraints are added during post-processing right, so this would not help here. We need it to complete symex, convert SSA and then post process as well before terminating.

Reply:
True we would need to introduce an intermediate pipeline that stops later. Not sure whether this is worth doing...

Reply:
This might be useful for `--write-solver-stats-to` as well. When you have very slow proofs, you may not want to wait to get solver query statistics. I'm assuming here that all relevant info is already collected before the call to the solver.